File: go/middleware/guardrails/error_tracker.go
package guardrails

import "time"

type ErrorCategory string

const (
	ErrCategoryTimeout     ErrorCategory = "timeout"
	ErrCategoryNotFound    ErrorCategory = "not_found"
	ErrCategoryInvalidArgs ErrorCategory = "invalid_args"
	ErrCategoryPermission  ErrorCategory = "permission"
	ErrCategoryRateLimit   ErrorCategory = "rate_limit"
	ErrCategoryUnknown     ErrorCategory = "unknown"
)

type ToolError struct {
	Timestamp  time.Time
	Tool       string
	Args       map[string]interface{}
	Category   ErrorCategory
	Message    string
	SessionID  string
	RetryCount int
}

type ErrorTracker struct {
	errors           map[string][]ToolError
	maxErrorsPerTool int
	windowDuration   time.Duration
}

func NewErrorTracker() *ErrorTracker {
	return &ErrorTracker{
		errors:           make(map[string][]ToolError),
		maxErrorsPerTool: 5,
		windowDuration:   time.Minute * 5,
	}
}

func (et *ErrorTracker) SetThresholds(maxErrors int, window time.Duration) {
	et.maxErrorsPerTool = maxErrors
	et.windowDuration = window
}

func (et *ErrorTracker) RecordError(sessionID, tool string, args map[string]interface{}, err error, category ErrorCategory) {
	if et.errors[sessionID] == nil {
		et.errors[sessionID] = make([]ToolError, 0)
	}

	toolErr := ToolError{
		Timestamp:  time.Now(),
		Tool:       tool,
		Args:       args,
		Category:   category,
		Message:    err.Error(),
		SessionID:  sessionID,
		RetryCount: et.getRetryCount(sessionID, tool),
	}

	et.errors[sessionID] = append(et.errors[sessionID], toolErr)
	et.pruneOldErrors(sessionID)
}

func (et *ErrorTracker) getRetryCount(sessionID, tool string) int {
	count := 0
	for _, err := range et.errors[sessionID] {
		if err.Tool == tool {
			count++
		}
	}
	return count
}

func (et *ErrorTracker) pruneOldErrors(sessionID string) {
	if len(et.errors[sessionID]) <= et.maxErrorsPerTool {
		return
	}

	cutoff := time.Now().Add(-et.windowDuration)
	var filtered []ToolError
	for _, err := range et.errors[sessionID] {
		if err.Timestamp.After(cutoff) {
			filtered = append(filtered, err)
		}
	}
	et.errors[sessionID] = filtered
}

func (et *ErrorTracker) GetErrorCount(sessionID, tool string) int {
	count := 0
	for _, err := range et.errors[sessionID] {
		if err.Tool == tool {
			count++
		}
	}
	return count
}

func (et *ErrorTracker) GetRecentErrors(sessionID string) []ToolError {
	et.pruneOldErrors(sessionID)
	result := make([]ToolError, len(et.errors[sessionID]))
	copy(result, et.errors[sessionID])
	return result
}

func (et *ErrorTracker) GetErrorsByCategory(sessionID string, category ErrorCategory) []ToolError {
	var result []ToolError
	for _, err := range et.errors[sessionID] {
		if err.Category == category {
			result = append(result, err)
		}
	}
	return result
}

func (et *ErrorTracker) IsErrorRateExceeded(sessionID, tool string) bool {
	count := et.GetErrorCount(sessionID, tool)
	return count >= et.maxErrorsPerTool
}

func (et *ErrorTracker) ResetSession(sessionID string) {
	delete(et.errors, sessionID)
}

func (et *ErrorTracker) ShouldBlockTool(sessionID, tool string) bool {
	return et.IsErrorRateExceeded(sessionID, tool)
}

File: typescript/src/middleware/guardrails/error_tracker.ts
export enum ErrorCategory {
  TIMEOUT = "timeout",
  NOT_FOUND = "not_found",
  INVALID_ARGS = "invalid_args",
  PERMISSION = "permission",
  RATE_LIMIT = "rate_limit",
  UNKNOWN = "unknown",
}

export interface ToolError {
  timestamp: Date;
  tool: string;
  args: Record<string, unknown>;
  category: ErrorCategory;
  message: string;
  sessionId: string;
  retryCount: number;
}

export class ErrorTracker {
  private errors: Map<string, ToolError[]>;
  private maxErrorsPerTool: number;
  private windowDurationMs: number;

  constructor(maxErrorsPerTool: number = 5, windowDurationMinutes: number = 5) {
    this.errors = new Map();
    this.maxErrorsPerTool = maxErrorsPerTool;
    this.windowDurationMs = windowDurationMinutes * 60 * 1000;
  }

  setThresholds(maxErrors: number, windowMinutes: number): void {
    this.maxErrorsPerTool = maxErrors;
    this.windowDurationMs = windowMinutes * 60 * 1000;
  }

  recordError(
    sessionId: string,
    tool: string,
    args: Record<string, unknown>,
    err: Error,
    category: ErrorCategory,
  ): void {
    if (!this.errors.has(sessionId)) {
      this.errors.set(sessionId, []);
    }

    const retryCount = this.getRetryCount(sessionId, tool);

    const toolErr: ToolError = {
      timestamp: new Date(),
      tool,
      args,
      category,
      message: err.message,
      sessionId,
      retryCount,
    };

    this.errors.get(sessionId)!.push(toolErr);
    this.pruneOldErrors(sessionId);
  }

  private getRetryCount(sessionId: string, tool: string): number {
    return (this.errors.get(sessionId) || []).filter((e) => e.tool === tool).length;
  }

  private pruneOldErrors(sessionId: string): void {
    const errors = this.errors.get(sessionId);
    if (!errors || errors.length <= this.maxErrorsPerTool) {
      return;
    }

    const cutoff = new Date(Date.now() - this.windowDurationMs);
    this.errors.set(
      sessionId,
      errors.filter((e) => e.timestamp > cutoff),
    );
  }

  getErrorCount(sessionId: string, tool: string): number {
    return (this.errors.get(sessionId) || []).filter((e) => e.tool === tool).length;
  }

  getRecentErrors(sessionId: string): ToolError[] {
    this.pruneOldErrors(sessionId);
    return [...(this.errors.get(sessionId) || [])];
  }

  getErrorsByCategory(sessionId: string, category: ErrorCategory): ToolError[] {
    return (this.errors.get(sessionId) || []).filter((e) => e.category === category);
  }

  isErrorRateExceeded(sessionId: string, tool: string): boolean {
    return this.getErrorCount(sessionId, tool) >= this.maxErrorsPerTool;
  }

  resetSession(sessionId: string): void {
    this.errors.delete(sessionId);
  }

  shouldBlockTool(sessionId: string, tool: string): boolean {
    return this.isErrorRateExceeded(sessionId, tool);
  }
}